Add clipboard support

ForTheReallys opened the pull request:
Tested with `gamescope gvim`.  This PR adds support for copying/pasting utf-8 text for both the clipboard and the primary selection.  In the event that SDL_CLIPBOARDUPDATE doesn't get triggered (for example, when running gamescope as XWayland under sway), I've implemented the keybinding `Super + C` to sync both the clipboard and primary selection with gamescope.  As of now, this PR only supports Xwayland application windows under gamescope.  I don't have much experience with Xlib, so any feedback is appreciated.

Closes #303

Joshua-Ashton commented on README.md:
```diff
@@ -39,6 +39,8 @@ meson install -C build/ --skip-subprojects
 * **Super + O** : Decrease FSR sharpness by 1
 * **Super + S** : Take screenshot (currently goes to `/tmp/gamescope_$DATE.png`)
+* **Super + G** : Toggle keyboard grab
+* **Super + S** : Update clipboard
```
C

emersion commented:
>  In the event that SDL_CLIPBOARDUPDATE doesn't get triggered (for example, when running gamescope as XWayland under sway), I've implemented the keybinding Super + C to sync both the clipboard and primary selection with gamescope.

That sounds like a bug somewhere. Sway should update the X11 clipboard when an X11 window is focused.

ForTheReallys commented:
@emersion Just found out it's a bug in SDL.  Looks SDL doesn't send CLIPBOARDUPDATE if the videodriver is x11.

Joshua-Ashton commented:
We can just fix it in SDL maybe?

ForTheReallys commented:
Works for me.  I'll look into that.  In the meantime, is there anything stopping us from merging this PR?

ForTheReallys commented:
SDL PR: https://github.com/libsdl-org/SDL/pull/7951

ForTheReallys commented:
SDL PR: https://github.com/libsdl-org/SDL/pull/7951

ForTheReallys commented:
The fix has been merged into SDL2.  Let me know if there's anything else you need from me

ForTheReallys commented:
@emersion @Joshua-Ashton Bump

Joshua-Ashton commented:
On Holiday until Wednesday

ForTheReallys commented:
Any updates on this @Joshua-Ashton?

ForTheReallys commented:
> With the SDL bug fixed, do we really need the keyboard shortcut to synchronize?

Really your call. I don't mind either way. I'm not super familiar with distro release schedules, but the only reason I can think of to keep it around is if this change gets picked up by distros before the SDL fix. Again, your call. Let me know and I'll do what I need to do

ForTheReallys commented:
@emersion I went ahead and removed the keybind now that SDL has been fixed to properly report clipboard updates.  I also added support for clipboard synchronization across xwayland instances in embedded mode.  I was going to wait until this PR got merged, but it was a pretty simple change, so I just included it in this PR.  That commit should also fix #916.

Joshua-Ashton commented on src/steamcompmgr.cpp:
```diff
@@ -4611,14 +4609,28 @@ void x11_set_selection(xwayland_ctx_t *ctx, std::string contents, int selectionT
 
+	for (int i = 0; i < g_nXWaylandCount; i++)
+	{
+		gamescope_xwayland_server_t *server = wlserver_get_xwayland_server(i);
```
Should call this like the other things that loop with `gamescope_xwayland_server_t` and also avoid updating the atom for the server that's calling it as we don't want to end up in a race with buffered messages on the same server.

ForTheReallys replied:
Can you elaborate on what you mean by "like the other things that loop with `gamescope_xwayland_server_t`?"

Joshua-Ashton replied:
```
		gamescope_xwayland_server_t *server = NULL;
		for (size_t i = 0; (server = wlserver_get_xwayland_server(i)); i++)
```

ForTheReallys replied:
As it turns out, we shouldn't actually need to update the atom at all with this call in any of the instances, since that gets done when we receive a SelectionRequest.  All we really want to do here is become the selection owner.  Again, I'm not all that familiar with Xlib, so please let me know if I'm missing something

ForTheReallys commented:
Bump @Joshua-Ashton

ForTheReallys commented:
Bump @Joshua-Ashton

Joshua-Ashton commented:
Thought I merged this already, oops!

ForTheReallys commented:
No worries.  Thank you!

Preposterone commented:
This PR has been merged, however, I don't see this functionality neither in readme of current codebase, nor in help message of a built-from source executable.

Has this been reverted for some reason? If not, what are the steps to enable basic clipboard functionality?

@ForTheReallys @Joshua-Ashton

ForTheReallys commented:
@Preposterone the easiest way to get this working is to set `SDL_VIDEODRIVER=wayland`  (assuming you're on Wayland). For it to work with X11 or XWayland, you'll need to have SDL 2.30.0.

MisterrrX commented:
> @Preposterone the easiest way to get this working is to set `SDL_VIDEODRIVER=wayland` (assuming you're on Wayland). For it to work with X11 or XWayland, you'll need to have SDL 2.30.0.

Unfortunately, this doesn't work for me on Arch and Hyprland. Has anyone got this to work?

mikeyjoel commented:
I play FFXVI often and need to login using my SquareEnix Mogstation provided credentials. I can't copy and paste from 1Password onto it. Tried some a different app, doesnt work. Disabling gamescope allows for clipboard functionality. Using Tumbleweed with Steam installed from the games repo via zypper natively.
